controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -944,118 +944,158 @@ private boolean isSubscriberToNonSubscriberTransition(final ReaderGroupConfigRec
+                                                 }
+                                             });
+                                   })).orElse(CompletableFuture.completedFuture(null));
+                    } else {
                         // if retained size is less than max allowed, then truncate the stream at subscriber lower bound.
-                        if (retainedSize < policy.getRetentionMax()) {
+                        if (retainedSizeLB < policy.getRetentionMax()) {
                             return CompletableFuture.completedFuture(lowerBound);
-                        } else {
-                            // if retained size is greater than max allowed, then truncate the stream at streamcut
-                            // from retention set that matches the retention policy size bound.
-                            return maxBound.get().map(x -> streamMetadataStore.getStreamCutRecord(scope, stream, x, context, executor)
-                            .thenApply(StreamCutRecord::getStreamCut)
-                            .thenCompose(maxRecord -> {
-                            // if max record is strictly greater than lowerbound then we can truncate at max record
-                            return streamMetadataStore.streamCutStrictlyGreaterThan(
-                            scope, stream, maxRecord, lowerBound, context, executor)
-                            .thenApply(gt -> {
-                            if (gt) {
-                                return maxRecord;
-                              } else {
-                                return lowerBound;
-                              }
-                            });
-                            })).orElse(CompletableFuture.completedFuture(null));
+                        } else { // greater than max. truncate at max. 
+                            // let there be data loss. its a lagging reader.. but if there is no streamcut in 
+                            // retentionset that satisfied the min and max criteria, then we should simply truncate at 
+                            // least at the lowerbound 
+                            return Optional.ofNullable(limits.getKey())
+                                           .filter(x -> currentSize - x.getRecordingSize() < retainedSizeLB)
+                                           .map(x -> streamMetadataStore.getStreamCutRecord(scope, stream, x, context, executor)
+                                                                        .thenApply(StreamCutRecord::getStreamCut))
+                                    .orElse(CompletableFuture.completedFuture(lowerBound));
                         }
-                      }
+                    }
                 });
     }
 
     private CompletableFuture<Map<Long, Long>> getTruncationStreamCutByTimeLimit(String scope, String stream, OperationContext context,
```

Comment:
Inside this method it may be good to explain the flow of code at significant points. The logic involved here is very complex and very difficult to understand/recollect without any comments in the code.

Reply:
i looked at the comments.. and i see comments describing each step in the code along the way.. i understand that some of it may not be clear.. but is there any part specifically that do you want described with more clarity.